ghcide/src/Development/IDE/Types/Logger.hs

```diff
@@ -57,3 +91,180 @@ logTelemetry x = logPriority x Telemetry
 
 noLogging :: Logger
 noLogging = Logger $ \_ _ -> return ()
+
+data WithPriority a = WithPriority { priority :: Priority, payload :: a } deriving Functor
+
+-- | Note that this is logging actions _of the program_, not of the user.
+--   You shouldn't call warning/error if the user has caused an error, only
+--   if our code has gone wrong and is itself erroneous (e.g. we threw an exception).
+data Recorder msg = Recorder
+  { logger_ :: forall m. (HasCallStack, MonadIO m) => msg -> m () }
+
+logWith :: (HasCallStack, MonadIO m) => Recorder (WithPriority msg) -> Priority -> msg -> m ()
+logWith recorder priority msg = withFrozenCallStack $ logger_ recorder (WithPriority priority msg)
```

Comment:
If you switch the order of the constructors of `WithPriority` then you could just make this
```
logWith :: ... => Recorder msg -> msg -> m ()
```
and have the call sites do this
```
logWith logger $ "hello" `WithPriority` Info
```
which is a bit more verbose but kind of cute :) Probably this is better, given this gets called a lot.

Reply:
It is cooler to have `logWith` not know anything about `Priority`. I think in the end going back to
```
logDebug logger log = logWith logger $ WithPriority Debug log
```
will be better since I expect everything to use priorities/severities